### PR TITLE
Datastore API module refactor

### PR DESCRIPTION
**Pull request by topicus**
Issue: CIVIC-5324, CIVIC-5573

### Description
We experienced several issues trying to make the dkan_datastore_api module satisfy all the react-dashboard requirements. So we decided to refactor the module to make it easy to follow the code and at the same time fix issues.

- [ ] Rename functions and unify names to improve legibility. There were several function names inherited from previous implementations that doesn't make sense anymore. We created one function that map to each SQL operation so there is a dkan_datastore_api_where, dkan_datastore_api_join, dkan_datastore_api_sort, etc.
- [ ] Fix an issue with joins that was causing fields being added multiple times in the result set. That also impact in the resulting performance.
- [ ] Move datastore api documentation to read the docs.
- [ ] Creates a cache layer to the datastore api. 
- [ ] Add the ability to aggregate multiple fields in the same query (e.g. `{"sum": ["County1","County2"]}`.
- [ ] Update tests to use the latest function names.

### QA steps for aggregations
- [ ] Extract the csv from the resource attached, upload and import to the datastore.
[physician_data_test_1 (2).csv.zip](https://github.com/NuCivic/dkan/files/755906/physician_data_test_1.2.csv.zip)
- [ ] Using postman or similar post to `/api/action/datastore/search.json` with the following signatures:

One field:
```
{
	"by_county": {
		"group_by": {"mytable":"WL_1_County"},
		"sum": "County1",
		"resource_id": {"mytable" : "REPLACE WITH THE UUID OF THE RESOURCE"},
		"limit": 10000000,
		"filters": {}
	}
}
```
Multiple fields:
```
{
	"by_county": {
		"group_by": {"mytable":"WL_1_County"},
		"sum": ["County1", "County2"],
		"resource_id": {"mytable" : "REPLACE WITH THE UUID OF THE RESOURCE"},
		"limit": 10000000,
		"filters": {}
	}
}
```
Multiple fields with table prefix:
```
{
	"by_county": {
		"group_by": {"mytable":"WL_1_County"},
		"sum": {"mytable": {"County1": "MYCOUNTY", "County2": "MYCOUNTY2"}},
		"resource_id": {"mytable" : "REPLACE WITH THE UUID OF THE RESOURCE"},
		"limit": 10000000,
		"filters": {}
	}
}
```

### Acceptance criteria
- [ ] Tests should pass
- [ ] No errors should be thrown
- [ ] Aggregate fields should appear in results accordingly passed params

### CODE REVIEW

**Comment by starsinmypockets** [comment]
@dafeder @topicus Let me test this against the gbpw project before merging.

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module [hunk 1]
```diff
@@ -360,7 +360,12 @@ function dkan_datastore_api_services_resources() {
 function _dkan_datastore_api_datastore_index() {
   if ($_SERVER['REQUEST_METHOD'] === 'GET') {
     $params = _dkan_datastore_api_get_params();
+    $cache_hash = md5(json_encode($params));
```
Comment (dafeder) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: @topicus can you document this function more completely? What kind of index is being created, and explain that this adds responses to the cache.

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module [hunk 2]
```diff
@@ -575,15 +585,43 @@ function dkan_datastore_api_add_index_group_by(&$data_select, $group_by, $alias
 /**
  * Adds aggregations to query.
  */
-function dkan_datastore_api_add_index_aggregations(&$data_select, $aggregations, $alias = NULL) {
+function dkan_datastore_api_add_index_aggregations(&$data_select, $aggregations) {
+
+  // { "sum": {"table_name": {"field_name": "as_field" }}}
```
Comment (dafeder) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: These comments aren't readable. Replace with brief explanations of what's going on at each stuff.

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/README.md [hunk 3]
```diff
@@ -229,3 +229,9 @@ http://example.com/api/dataset/search?resource_id[pop]=d3c099c6-1340-4ee5-b030-8
 
 Returns the country, population, squarekm and id for US and AR from datasets 11 and 13.
 
+### Caching
```
Comment (dafeder) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/README.md]: Now that RTD is live, move this to the API documentation in /docs please!

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module (hunk 1 above)
Comment (topicus) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: @dafeder names are legacy. Remember this is not a completely new implementation but a bypass between the existing code and what we needed for dashboards. 
https://github.com/NuCivic/dkan_datastore/blob/cd5cd595f8fe11f7289cbf1073a1c8f191f2f9e5/modules/dkan_datastore_api/dkan_datastore_api.module#L87

There is a lot of names that need to be changed to reflect the right behavior.  

I can work on renaming this if you want.

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module (hunk 2 above)
Comment (topicus) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: @dafeder Each comment added that it's an object shows what's the case that branch of the code it's handling. Maybe `Handle params formatted as: { "sum": {"table_name": {"field_name": "as_field" }}}`?

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/README.md (hunk 3 above)
Comment (topicus) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/README.md]: Yes, didn't know this was live, I'll document there from now. Is it a dev or end user doc?

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module (hunk 2 above)
Comment (dafeder) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: Yes, that's what I meant :)

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module [hunk 4]
```diff
@@ -533,21 +547,25 @@ function dkan_datastore_api_add_index_conditions(&$data_select, $filters) {
...
           $value = is_array($value) ? $value : explode(',', $value);
-          $data_select->condition($key, $value, 'IN');
+          $data_select->condition($key, $value, dkan_datastore_api_clause_condition($value));
         }
       }
     }
   }
 }
 
+function dkan_datastore_api_clause_condition($value) {
```
Comment (topicus) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: Document this.

File: modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module [hunk 5]
```diff
@@ -722,36 +781,55 @@ function dkan_datastore_api_build_index_query($table, $offset, $limit, $fields,
...
-function dkan_datastore_api_resource_build_index_query(&$query, $page, $fields, $page_size, $alias, $table) {
-  $default_limit = variable_get('dkan_datastore_default_page_size', 100);
-  if ((!user_access('perform unlimited index queries') && $page_size > $default_limit) || !$page_size) {
-    $page_size = $default_limit;
+function dkan_datastore_api_select_fields(&$data_select, $fields, $resource_ids) {
+  $resource_ids = is_assoc($resource_ids)
+    ? $resource_ids
+    : array( DKAN_DATASTORE_API_DEFAULT_TABLE_ALIAS => $resource_ids[0]);
+
+  // Select all the fields using an alias performs better than selecting
```
Comment (topicus) [inline modules/dkan/dkan_datastore/modules/dkan_datastore_api/dkan_datastore_api.module]: Document performance difference between both approaches.

**Pull request merged**